rmw_opendds_cpp/src/rmw_take.cpp

```diff
@@ -26,179 +26,114 @@
-  if (!data_reader) {
+  RMW_CHECK_FOR_NULL_WITH_MSG(subscription, "subscription is null", return RMW_RET_ERROR);
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(subscription handle,
+    subscription->implementation_identifier, opendds_identifier, return RMW_RET_ERROR)
+
+  auto info = static_cast<OpenDDSStaticSubscriberInfo *>(subscription->data);
+  RMW_CHECK_FOR_NULL_WITH_MSG(info, "subscriber info is null", return RMW_RET_ERROR);
+  RMW_CHECK_FOR_NULL_WITH_MSG(info->topic_reader_, "topic_reader_ is null", return RMW_RET_ERROR);
+  RMW_CHECK_FOR_NULL_WITH_MSG(info->callbacks_, "callbacks_ is null", return RMW_RET_ERROR);
+
+  RMW_CHECK_FOR_NULL_WITH_MSG(cdr_stream, "cdr_stream is null", return RMW_RET_ERROR);
+  RMW_CHECK_FOR_NULL_WITH_MSG(taken, "taken is null", return RMW_RET_ERROR);
+  *taken = false;
+
+  OpenDDSStaticSerializedDataDataReader_var reader =
+    OpenDDSStaticSerializedDataDataReader::_narrow(info->topic_reader_);
+  if (!reader) {
     RMW_SET_ERROR_MSG("failed to narrow data reader");
-    return false;
+    return RMW_RET_ERROR;
   }
 
   OpenDDSStaticSerializedDataSeq dds_messages;
   DDS::SampleInfoSeq sample_infos;
-  bool ignore_sample = false;
-
-  DDS::ReturnCode_t status = data_reader->take(
-    dds_messages,
-    sample_infos,
-    1,
-    DDS::ANY_SAMPLE_STATE,
-    DDS::ANY_VIEW_STATE,
-    DDS::ANY_INSTANCE_STATE);
-  if (status == DDS::RETCODE_NO_DATA) {
-    data_reader->return_loan(dds_messages, sample_infos);
-    *taken = false;
-    return true;
-  }
-  if (status != DDS::RETCODE_OK) {
-    RMW_SET_ERROR_MSG("take failed");
-    data_reader->return_loan(dds_messages, sample_infos);
-    return false;
-  }
-
-  DDS::SampleInfo & sample_info = sample_infos[0];
-  if (!sample_info.valid_data) {
-    // skip sample without data
-    ignore_sample = true;
-  } else if (ignore_local_publications) {
-    // compare the lower 12 octets of the guids from the sender and this receiver
-    // if they are equal the sample has been sent from this process and should be ignored
-    //DDS::GUID_t sender_guid = sample_info.original_publication_virtual_guid;
-    DDS::InstanceHandle_t receiver_instance_handle = dds_data_reader->get_instance_handle();
-    ignore_sample = true;
-    for (size_t i = 0; i < 12; ++i) {
-      //CORBA::Octet * sender_element = &(sender_guid.value[i]);
-      CORBA::Octet * receiver_element =
-        &(reinterpret_cast<CORBA::Octet *>(&receiver_instance_handle)[i]);
-      //if (*sender_element != *receiver_element) {
-      //  ignore_sample = false;
-      //  break;
-      //}
+  DDS::ReturnCode_t status = reader->take(dds_messages, sample_infos, 1,
+    DDS::ANY_SAMPLE_STATE, DDS::ANY_VIEW_STATE, DDS::ANY_INSTANCE_STATE);
+
+  if (DDS::RETCODE_OK == status) {
+    DDS::SampleInfo & info = sample_infos[0];
+    if (info.valid_data) {
+      cdr_stream->buffer_length = dds_messages[0].serialized_data.length();
+      if (cdr_stream->buffer_length <= (std::numeric_limits<unsigned int>::max)()) {
```

Comment:
The parentheses around `std::numeric_limits<unsigned int>::max` look look strange. May just remove them.

Reply:
It is a pattern used in some projects for compatibility with systems that have max defined as a function-like macro